src/realm/object-store/c_api/util.cpp

```diff
@@ -3,20 +3,18 @@
 
 namespace realm::c_api {
 
-template <class T>
-inline T* cast_ptr(void* ptr)
+inline WrapC* cast_ptr(void* ptr)
 {
     auto rptr = static_cast<WrapC*>(ptr);
-    REALM_ASSERT(dynamic_cast<T*>(rptr) != nullptr);
-    return static_cast<T*>(rptr);
+    REALM_ASSERT(rptr->cookie == WrapC::s_cookie_value);
```

Comment:
do we also want to REALM_ASSERT(dynamic_cast<>) here? it seems like this will check that the ptr hasn't been destroyed or corrupted, but doesn't check that the type is correct. Did this change come from some failure investigation where checking the type wasn't important but checking that the ptr hasn't been corrupted was important?

Reply:
This came from a test where I tried to `realm_release` a pointer to something that was not a WrapC pointer and it was not caught by this test. When we have the initial static cast to WrapC, then we assume that rptr is indeed a WrapC pointer and the following dynamic cast has no effect as we only ever used this function to cast to WrapC. This is why I removed the template stuff.

Reply:
Gotcha. Makes sense now.